arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorizedArrowReader.java

```diff
@@ -0,0 +1,387 @@
+    this.isIntType = false;
+    this.isLongType = false;
+    this.isFloatType = false;
+    this.isDoubleType = false;
+    this.vectorizedColumnIterator = null;
+  }
+
+  @SuppressWarnings("checkstyle:CyclomaticComplexity")
+  @Override
+  public VectorHolder read(int numValsToRead) {
+    if (vec == null || !reuseContainers) {
+      typeWidth = allocateFieldVector();
+    }
+    vec.setValueCount(0);
+    nullabilityHolder = new NullabilityHolder(batchSize);
+    if (vectorizedColumnIterator.hasNext()) {
+      if (allPagesDictEncoded) {
+        vectorizedColumnIterator.nextBatchDictionaryIds((IntVector) vec, nullabilityHolder);
+      } else {
+        if (isFixedLengthDecimal) {
+          vectorizedColumnIterator.nextBatchFixedLengthDecimal(vec, typeWidth, nullabilityHolder);
+          ((IcebergDecimalArrowVector) vec).setNullabilityHolder(nullabilityHolder);
+        } else if (isFixedWidthBinary) {
+          // Fixed width binary type values are stored in an IcebergVarBinaryArrowVector as well
+          if (vec instanceof IcebergVarBinaryArrowVector) {
+            ((IcebergVarBinaryArrowVector) vec).setNullabilityHolder(nullabilityHolder);
+          }
+          vectorizedColumnIterator.nextBatchFixedWidthBinary(vec, typeWidth, nullabilityHolder);
+        } else if (isVarWidthType) {
+          if (vec instanceof IcebergVarcharArrowVector) {
+            ((IcebergVarcharArrowVector) vec).setNullabilityHolder(nullabilityHolder);
+          } else if (vec instanceof IcebergVarBinaryArrowVector) {
+            ((IcebergVarBinaryArrowVector) vec).setNullabilityHolder(nullabilityHolder);
+          }
+          vectorizedColumnIterator.nextBatchVarWidthType(vec, nullabilityHolder);
+        } else if (isBooleanType) {
+          vectorizedColumnIterator.nextBatchBoolean(vec, nullabilityHolder);
+        } else if (isPaddedDecimal) {
+          ((IcebergDecimalArrowVector) vec).setNullabilityHolder(nullabilityHolder);
+          vectorizedColumnIterator.nextBatchIntLongBackedDecimal(vec, typeWidth, nullabilityHolder);
+        } else if (isIntType) {
+          vectorizedColumnIterator.nextBatchIntegers(vec, typeWidth, nullabilityHolder);
+        } else if (isLongType) {
+          vectorizedColumnIterator.nextBatchLongs(vec, typeWidth, nullabilityHolder);
+        } else if (isFloatType) {
+          vectorizedColumnIterator.nextBatchFloats(vec, typeWidth, nullabilityHolder);
+        } else if (isDoubleType) {
+          vectorizedColumnIterator.nextBatchDoubles(vec, typeWidth, nullabilityHolder);
+        }
+      }
+    }
+    if (vec.getValueCount() != numValsToRead) {
+      throw new IllegalStateException("Number of values read into the vector, " +
+          vec.getValueCount() + " is not the same as the expected count of " + numValsToRead);
+    }
+    return new VectorHolder(columnDescriptor, vec, allPagesDictEncoded, dictionary, nullabilityHolder);
+  }
+
+  private int allocateFieldVector() {
```

Comment:
I'd normally expect an `allocateSomething` method to return the something, not a related int. Could this set the type width and return `void` instead? Or can the type width be derived from the vector that was allocated so this can return one thing?